### PR TITLE
gh-98586: Add vector call APIs to the Limited API

### LINKED ISSUE #98586: Add vector call functions to the limited API
# Feature or enhancement

Starting in Python 3.12, it is possible to _receive_ vector calls in the limited API. However, it is not possible to _issue_ them from a binary extension module compiled for the limited API.

# Pitch

Since the PEP-590 is presumably stable at this point, I propose exposing more of it through the limited API. This would entail the functions

- `PyObject_Vectorcall`
- `PyObject_VectorcallMethod`
- `PyObject_VectorcallDict`
- `PY_VECTORCALL_ARGUMENTS_OFFSET` (a macro constant that is needed to use these effectively)

# Previous discussion

Some related discussion took place here: https://discuss.python.org/t/ideas-for-forward-compatible-and-fast-extension-libraries-in-python-3-12. This PR Is specifically about item #5 from the list in the first post.

* PR: gh-99056

### PR DESCRIPTION
**Pull request by wjakob**
This PR constains tentative changes needed to expose the facilities for making PEP-590-style vector calls through Python's limited API.

* Issue: gh-98586

### CODE REVIEW

**Comment by erlend-aasland** [comment]
@markshannon might be interested in reviewing this as the author of PEP 590.

**Review by erlend-aasland — COMMENTED** [review]
Could you add tests to `Modules/_testcapimodule/vectorcall_limited.c` and also adjust the NEWS entry to use Sphinx directives for linking to the relevant parts of the C-API docs and PEP-590?

**Comment by erlend-aasland** [comment]
What is the reason for not adding `PyVectorcall_Function` to the Limited API?

**Comment by wjakob** [comment]
> What is the reason for not adding PyVectorcall_Function to the Limited API?

It seemed like sort of an implementation detail. I am happy to add it (& tests) if there is consensus to do so.

**Comment by erlend-aasland** [comment]
> It seemed like sort of an implementation detail. I am happy to add it (& tests) if there is consensus to do so.

I've never had use for it myself, so I have no opinion about this; I was just curious. Let's hear what the others say when they chime in.

File: Misc/NEWS.d/next/Core and Builtins/2022-10-24-10-30-30.gh-issue-98586.Tha5Iy.rst
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Core and Builtins/2022-10-24-10-30-30.gh-issue-98586.Tha5Iy.rst]:
```suggestion
the :ref:`Limited API <stable>` along with the auxiliary macro constant
```

File: Modules/_testcapi/vectorcall_limited.c
Comment (erlend-aasland) [inline Modules/_testcapi/vectorcall_limited.c]: Minor PEP 7 nit (ditto for the `if`s below): 

```suggestion
        PyTuple_SetItem(kwnames, 0, kwname))
    {
```

**Comment by markshannon** [comment]
Two questions.

##### First question
Why include `PyObject_VectorcallDict`? It seem to be an internal function for efficiently supporting the old calling convention for callables that support vectorcall.

#### Second question
The vectorcall API has two parts. 
1. Calling through `PyObject_Vectorcall` 
2. Declaring that an object can be called as "vector callable" by setting the `Py_TPFLAGS_HAVE_VECTORCALL` flag and `type.tp_vectorcall_offset`.

Any plans to add the ability to declare classes as having "vector callable" instances?

**Comment by wjakob** [comment]
Thanks for the review @markshannon.

> Why include PyObject_VectorcallDict? It seem to be an internal function for efficiently supporting the old calling convention for callables that support vectorcall.

I am happy to hear it. I have never had the need to use this function myself. I included in the PR for reasons of symmetry, and because it was prominently advertised in the docs next to `PyObject_Vectorcall` and `PyObject_VectorcallMethod`.

> Any plans to add the ability to declare classes as having "vector callable" instances?

I think that this was already done in PR https://github.com/python/cpython/issues/93274. Please let me know if I misunderstood.

Finally, do you have any thoughts on `PyVectorcall_Function` brought up by @erlend-aasland ?

**Comment by markshannon** [comment]
One other question.

Why add `PyObject_VectorcallMethod`, rather than exposing the lower level method-getting machinery?
`PyObject_VectorcallMethod` forces the attribute lookup to happen after the arguments are evaluated which is different from Python.
A `PyObject *LoadMethod(PyObject *obj, PyObject *name, PyObject **self)` function would allow C code to use Python evaluation order. `PyObject_VectorcallMethod` could be added as a utility function if desirable.

This is a sketch and is probably missing some error handling and/or incref/decrefs.
```C
PyObject *PyObject_VectorcallMethod(
    PyObject *name, PyObject *const *args,
    size_t nargsf, PyObject *kwnames)
{
      /* Error if nargs < 1 */
      PyObject *obj = args[0];
      PyObject *callable = LoadMethod(obj, name, &args[0]);
      /* TO DO --- Handle error if callable == NULL */
      if (args[0] == NULL) {
            /* No self */
           return PyObject_Vectorcall(callable, args+1, (nargsf-1) | PY_VECTORCALL_ARGUMENTS_OFFSET, kwnames);
      }
      return PyObject_Vectorcall(callable, args, nargsf, kwnames);
}
```

**Comment by markshannon** [comment]
> Finally, do you have any thoughts on PyVectorcall_Function brought up by @erlend-aasland ?

I'd rather not add it.
Unless it is specified in such a way as to be useless, its existence would unnecessarily constrain the implementation.

**Comment by wjakob** [comment]
> Why add `PyObject_VectorcallMethod`, rather than exposing the lower level method-getting machinery?
`PyObject_VectorcallMethod` forces the attribute lookup to happen after the arguments are evaluated which is different from Python.

It is convenient to use a 1-function-does-it-all interface like `PyObject_VectorcallMethod` for performing method calls from binding projects. This is particularly true for the `Py_LIMITED_API` that is already heavy on CPython API calls due to the opaque nature of the interfaces. The difference in attribute lookup ordering is an interesting point. I was not aware of this, but it strikes me as a quite minor/subtle issue. (Maybe others will disagree though!)

Altogether, it sounds to me more like a general question about why `PyObject_VectorcallMethod` was designed the way it was? My goal with this PR was to expose something that is already in general use.

**Comment by wjakob** [comment]
Update: removed `PyObject_VectorcallDict`, incorporated suggestions by @erlend-aasland.

**Comment by erlend-aasland** [comment]
> Update: removed `PyObject_VectorcallDict`

No, it's still added to the Limited API; you only removed the tests.

> incorporated suggestions by @erlend-aasland.

Thanks

**Comment by wjakob** [comment]
@erlend-aasland — Oops, you are right. Should be fixed now.

**Review by erlend-aasland — APPROVED** [review]
LGTM. Thanks! This is a very nice addition to the Limited API.

**Review by Cammiec91 — APPROVED** [review]

File: Modules/_testcapi/vectorcall_limited.c
Comment (markshannon) [inline Modules/_testcapi/vectorcall_limited.c:46]: You need to check for errors after every call. Otherwise the exception object could leak.

File: Modules/_testcapi/vectorcall_limited.c
Comment (markshannon) [inline Modules/_testcapi/vectorcall_limited.c:93]: ditto

**Review by markshannon — DISMISSED** [review]
A couple of potential memory leaks, otherwise LGTM.

**Comment by wjakob** [comment]
Is this okay @markshannon? (with a goto to avoid lots of `if` blocks with different numbers of `Py_DECREF`s)

I have made the requested changes; please review again

**Comment by wjakob** [comment]
The `test_asyncio` test fails on Windows (Azure Pipelines). Potentially a fluke/unrelated issue? I don't see how it could be related.

**Comment by AlexWaygood** [comment]
> The `test_asyncio` test fails on Windows (Azure Pipelines). Potentially a fluke/unrelated issue? I don't see how it could be related.

It's a known issue that will be fixed by #98704

File: Modules/_testcapi/vectorcall_limited.c
Comment (erlend-aasland) [inline Modules/_testcapi/vectorcall_limited.c]:
```suggestion
static PyObject *
call_vectorcall(PyObject *self, PyObject *callable)
{
```

File: Modules/_testcapi/vectorcall_limited.c
Comment (erlend-aasland) [inline Modules/_testcapi/vectorcall_limited.c]:
```suggestion
static PyObject *
call_vectorcall_method(PyObject *self, PyObject *callable)
{
```

File: Modules/_testcapi/vectorcall_limited.c
Comment (encukou) [inline Modules/_testcapi/vectorcall_limited.c]: The comment can go away now.

**Review by encukou — APPROVED** [review]
I also have a nitpick, otherwise LGTM.
I'll push a commit to your branch to avoid another round of review ping-pong.

**Comment by encukou** [comment]
Thank you for the PR!

**Comment by wjakob** [comment]
Awesome, many thanks for your help in getting it merged!

**Comment by wjakob** [comment]
I realized that I forgot to add an entry to `Doc/whatsnew/3.12.rst`. Should I open a separate PR, or will somebody from the maintainers team do this?

**Comment by markshannon** [comment]
Open a new PR please.

**Pull request merged**
